File: LeetCode/Python/Medium/516-Longest-Palindromic-Subsequence/solution.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        n=len(s)

        left=0

        right=n-1

        result=self.helper(s, left, right)

        return result

    def helper(self, s: str, left: int, right: int) -> int:
        if left==right:
            return 1
        
        if left > right:
            return 0

        subproblem_solution=0

        if s[left]==s[right]:
            subproblem_solution=2 + self.helper(s, left + 1, right - 1)
        else:
            pick_left_character=self.helper(s, left, right - 1)

            pick_right_character=self.helper(s, left + 1, right)

            subproblem_solution=max(pick_left_character, pick_right_character)
        
        return subproblem_solution

class Solution:
    def __init__(self):
        self.memo=list()

    def longestPalindromeSubseq(self, s: str) -> int:
        n=len(s)

        self.memo=[[-1 for _ in range(0, n)] for _ in range(0, n)]

        left=0

        right=n-1

        result=self.helper(s, left, right)

        return result

    def helper(self, s: str, left: int, right: int) -> int:
        if left==right:
            return 1
        
        if left > right:
            return 0

        if self.memo[left][right]!=-1:
            return self.memo[left][right]

        subproblem_solution=0

        if s[left]==s[right]:
            subproblem_solution=2 + self.helper(s, left + 1, right - 1)
        else:
            pick_left_character=self.helper(s, left, right - 1)

            pick_right_character=self.helper(s, left + 1, right)

            subproblem_solution=max(pick_left_character, pick_right_character)

        self.memo[left][right]=subproblem_solution
        
        return subproblem_solution

File: LeetCode/Rust/Medium/516-Longest-Palindromic-Subsequence/solution.rs
impl Solution
{
    pub fn longest_palindrome_subseq(s: String) -> i32
    {
        let characters: Vec<char>=s.chars().collect::<Vec<char>>();

        let n: usize=characters.len();

        let mut memo: Vec<Vec<i32>>=vec![vec![-1; n]; n];

        let mut left: usize=0;

        let mut right: usize=n - 1;

        let result: i32=Self::helper(&mut memo, &characters, left, right);

        result
    }

    pub fn helper(memo: &mut Vec<Vec<i32>>, characters: &Vec<char>, left: usize, right: usize) -> i32
    {
        if left==right
        {
            return 1;
        }

        if left > right
        {
            return 0;
        }

        if memo[left as usize][right as usize]!=-1
        {
            return memo[left as usize][right as usize];
        }

        let mut subproblem_solution: i32=0;

        if characters[left as usize]==characters[right as usize]
        {
            subproblem_solution=2 + Self::helper(memo, characters, left + 1, right - 1);
        }
        else
        {
            let pick_left_character: i32=Self::helper(memo, characters, left, right - 1);

            let pick_right_character: i32=Self::helper(memo, characters, left + 1, right);

            subproblem_solution=i32::max(pick_left_character, pick_right_character);
        }

        memo[left as usize][right as usize]=subproblem_solution;

        subproblem_solution
    }
}